src/Dictionaries/IDictionary.cpp

```diff
@@ -0,0 +1,154 @@
+            "Dictionary {} key lookup structure does not match, expected {}",
+            getFullName(),
+            dictionary_structure.getKeyDescription());
+
+    for (size_t key_attribute_type_index = 0; key_attribute_type_index < key_attributes_types_size; ++key_attribute_type_index)
+    {
+        const auto & key_attribute_type = key_attributes_types[key_attribute_type_index];
+        auto & key_type = key_types[key_attribute_type_index];
+
+        if (key_attribute_type->equals(*key_type))
+            continue;
+
+        auto & key_column_to_cast = key_columns[key_attribute_type_index];
+        ColumnWithTypeAndName column_to_cast = {key_column_to_cast, key_type, ""};
+        auto casted_column = castColumnAccurate(column_to_cast, key_attribute_type);
+        key_column_to_cast = std::move(casted_column);
+        key_type = key_attribute_type;
+    }
+}
+
+Columns IDictionary::getColumns(
+    const Strings & attribute_names,
+    const DataTypes & result_types,
+    const Columns & key_columns,
+    const DataTypes & key_types,
+    const Columns & default_values_columns) const
+{
+    size_t attribute_names_size = attribute_names.size();
+
+    Columns result;
+    result.reserve(attribute_names_size);
+
+    for (size_t i = 0; i < attribute_names_size; ++i)
+    {
+        const auto & attribute_name = attribute_names[i];
+        const auto & result_type = result_types[i];
+        const auto & default_values_column = default_values_columns[i];
+
+        result.emplace_back(getColumn(attribute_name, result_type, key_columns, key_types, default_values_column));
+    }
+
+    return result;
+}
+
+Names IDictionary::getPrimaryKey() const
+{
+    return getStructure().getKeysNames();
+}
+
+Block IDictionary::getColumns(const Names & attribute_names, const ColumnsWithTypeAndName & key_columns, PaddedPODArray<UInt8> & found_keys_map) const
+{
+    size_t attribute_names_size = attribute_names.size();
+
+    Strings required_attributes_names;
+    DataTypes attributes_types;
+    Columns attributes_default_columns;
+    Block attributes_header;
+
+    const auto & dictionary_structure = getStructure();
+    size_t key_columns_size = key_columns.size();
+
+    if (attribute_names.empty())
+    {
+        size_t dictionary_attributes_size = dictionary_structure.attributes.size();
+        attributes_types.reserve(dictionary_attributes_size);
+        attributes_default_columns.reserve(dictionary_attributes_size);
+        required_attributes_names.reserve(dictionary_attributes_size);
+
+        /// If attribute names is empty, then use all attributes from dictionary structure
+        for (const auto & attribute : dictionary_structure.attributes)
+        {
+            attributes_types.emplace_back(attribute.type);
+            required_attributes_names.emplace_back(attribute.name);
+            attributes_default_columns.emplace_back(attribute.type->createColumnConst(key_columns_size, attribute.null_value));
```

Comment:
Before we did not support dictionary default value.
```
/// Dictinonary may have non-standart default values specified
default_cols[i] = result_types[i]->createColumnConstWithDefaultValue(out_null_map.size());
```
Probably it make sense to support it.

Reply:
It can be unexpected that this `getColumns` return defaults for type, while another overloads of `getColumns` return values from default columns (values specified by user in `CREATE DICTIONARY`)